pkg/controller/migplan/validation.go

```diff
@@ -329,3 +341,111 @@ func (r ReconcileMigPlan) validateDestinationCluster(plan *migapi.MigPlan) (int,
 
 	return 0, nil
 }
+
+// Validate required namespaces.
+// Returns error and the total error conditions set.
+func (r ReconcileMigPlan) validateRequiredNamespaces(plan *migapi.MigPlan) (int, error) {
+	totalSet := 0
+
+	// Source
+	nSet, err := r.validateSourceNamespaces(plan)
+	if err != nil {
+		return 0, err
+	}
+	totalSet += nSet
+
+	// Destination
+	nSet, err = r.validateDestinationNamespaces(plan)
+	if err != nil {
+		return 0, err
+	}
+	totalSet += nSet
+
+	return totalSet, nil
+}
+
+// Validate required namespaces on the source cluster.
+// Returns error and the total error conditions set.
+func (r ReconcileMigPlan) validateSourceNamespaces(plan *migapi.MigPlan) (int, error) {
+	namespaces := []string{velerorunner.VeleroNamespace}
+	cluster, err := plan.GetSourceCluster(r)
+	if err != nil {
+		return 0, err
+	}
+	if cluster == nil {
+		return 0, nil
+	}
+	client, err := cluster.GetClient(r)
+	if err != nil {
+		return 0, err
+	}
+	ns := kapi.Namespace{}
+	notFound := make([]string, 0)
+	for _, name := range namespaces {
+		key := types.NamespacedName{Name: name}
+		err := client.Get(context.TODO(), key, &ns)
+		if err == nil {
+			continue
+		}
+		if errors.IsNotFound(err) {
+			notFound = append(notFound, name)
+		} else {
+			return 0, err
+		}
+	}
+	if len(notFound) > 0 {
+		message := fmt.Sprintf(NsNotFoundOnSourceClusterMessage, strings.Join(notFound, ", "))
+		plan.Status.SetCondition(migapi.Condition{
+			Type:    NsNotFoundOnSourceCluster,
+			Status:  True,
+			Reason:  NotFound,
+			Message: message,
+		})
+		return 1, nil
+	}
+
+	return 0, nil
+}
+
+// Validate required namespaces on the destination cluster.
+// Returns error and the total error conditions set.
+func (r ReconcileMigPlan) validateDestinationNamespaces(plan *migapi.MigPlan) (int, error) {
```

Comment:
What are the "required namespaces" here? Velero will create any namespaces we are migrating as part of the "Restore" process.

Reply:
For both the source and destination, the required namespaces are not related to asset collection.  Instead, the list (to start with) includes the namespace the controller will create velero resources in.  Since it's intended to be configurable, It may not be `velero`.

Reply:
I see that now, thanks for the explanation.